jetbrains-ultimate/src/software/aws/toolkits/jetbrains/services/caws/DevEnvStatusWatcher.kt

```diff
@@ -0,0 +1,109 @@
+import software.aws.toolkits.resources.message
+
+class DevEnvStatusWatcher : StartupActivity {
+
+    companion object {
+        fun getInstance(project: Project) = project.service<DevEnvStatusWatcher>()
+        private val LOG = getLogger<DevEnvStatusWatcher>()
+    }
+
+    override fun runActivity(project: Project) {
+        if (!isRunningOnRemoteBackend()) {
+            return
+        }
+        val connection = SonoCredentialManager.getInstance(project).getConnectionSettings()
+            ?: error("Failed to fetch connection settings from Dev Environment")
+        val envId = System.getenv(CawsConstants.CAWS_ENV_ID_VAR) ?: error("envId env var null")
+        val org = System.getenv(CawsConstants.CAWS_ENV_ORG_NAME_VAR) ?: error("space env var null")
+        val projectName = System.getenv(CawsConstants.CAWS_ENV_PROJECT_NAME_VAR) ?: error("project env var null")
+        val client = connection.awsClient<CodeCatalystClient>()
+        val coroutineScope = projectCoroutineScope(project)
+        coroutineScope.launch(getCoroutineBgContext()) {
+            val initialEnv = client.getDevEnvironment {
+                it.id(envId)
+                it.spaceName(org)
+                it.projectName(projectName)
+            }
+            val inactivityTimeout = initialEnv.inactivityTimeoutMinutes()
+            if (inactivityTimeout == 0) {
+                LOG.info("Dev environment inactivity timeout is 0, not monitoring")
+                return@launch
+            }
+            val inactivityTimeoutInSeconds = inactivityTimeout * 60
+            var actualInactivityDuration: Long = 0
+            var lastControllerActivity: Long = 0
+            while (true) {
+                val statusJson = UnattendedStatusUtil.getStatus()
+                val lastActivityTime = statusJson.projects?.first()?.secondsSinceLastControllerActivity ?: 0
+                actualInactivityDuration = getActualInactivityDuration(lastActivityTime, lastControllerActivity)
+
+                if (actualInactivityDuration >= (inactivityTimeoutInSeconds - 300)) {
+                    try {
+                        val inactivityDurationInMinutes = actualInactivityDuration / 60
+                        val ans = runBlocking {
+                            val continueWorking = withContext(getCoroutineUiContext()) {
+                                return@withContext MessageDialogBuilder.okCancel(
+                                    message("caws.devenv.continue.working.after.timeout.title"),
+                                    message("caws.devenv.continue.working.after.timeout", inactivityDurationInMinutes)
+                                ).ask(project)
+                            }
+                            return@runBlocking continueWorking
+                        }
+
+                        if (ans) {
+                            notifyBackendOfActivity()
+                            lastControllerActivity = actualInactivityDuration
```

Comment:
won't this cause some desync since we're sending the environment API one time, but storing a slightly different one?

Reply:
Yeah, it may cause a couple of millisecond desync, let me change that to one value